activerecord/lib/arel.rb

```diff
@@ -28,28 +28,35 @@
 module Arel
   VERSION = "10.0.0"
 
-  # Wrap a known-safe SQL string for passing to query methods, e.g.
-  #
-  #   Post.order(Arel.sql("REPLACE(title, 'misc', 'zzzz') asc")).pluck(:id)
-  #
-  # Great caution should be taken to avoid SQL injection vulnerabilities.
-  # This method should not be used with unsafe values such as request
-  # parameters or model attributes.
-  def self.sql(raw_sql)
-    Arel::Nodes::SqlLiteral.new raw_sql
-  end
+  class << self
+    # Wrap a known-safe SQL string for passing to query methods, e.g.
+    #
+    #   Post.order(Arel.sql("REPLACE(title, 'misc', 'zzzz') asc")).pluck(:id)
+    #
+    # Wrap a known-save value to bypass sanitization, e.g.
+    #
+    #   Post.sanitize_sql_for_order([Arel.sql("field(id, ?)"), [Arel.sql(1)]])
+    #
+    # Great caution should be taken to avoid SQL injection vulnerabilities.
+    # This method should not be used with unsafe values such as request
+    # parameters or model attributes.
+    def sql(raw)
+      Arel::Nodes::SqlLiteral.new(raw.to_s)
```

Comment:
I share this reservation.

It's why I had two different implementations originally. In that the `Arel::SafeValue` returned the raw value, not the stringified version. Here because `class SqlLiteral < String` you get a raise:

```ruby
> Arel::Nodes::SqlLiteral.new(1)
`initialize': no implicit conversion of Integer into String (TypeError)
```

So a middle of the road compromise would be:

```ruby
    def sql(raw)
      Arel::Nodes::SqlLiteral.new(raw)
    end

    def safe(raw)
      Arel::Nodes::SqlLiteral.new(raw.to_s)
    end
```

But again that doesn't _really_ solve it.

This all originated from `Rational` not behaving as expected for `to_s`:

```ruby
> Rational(3, 2).to_s
"3/2"
```

So back to advocating for having a separate `Arel::SafeValue` class originally proposed in c717720ab25139ae927d6dd2c8df795924cd5b52

Reply:
Am I just confused? My expectation is that the result of `quote` is going to end up in an `#{}` interpolation in the middle of something that's definitely a string (of SQL).

That's also why I don't like the idea of the `safe` alias: SQL is SQL, and if the value you're passing isn't already a SQL string (or, newly introduced, a non-string that will `to_s` into something acceptably SQL-shaped), the caller is going to have to fix that in advance, regardless of how untouched we leave it.

@kmcphillips  can you help me understand how that original proposal helps Rational?

Reply:
The one danger I do see here is that we might end up encouraging people to do `Arel.sql(params[:limit])`, which they "know" is an integer, and we end up back in easily-written-injections land.

Considering that (and incidentally rejecting the idea [which hasn't yet come up, but feels like it's nearby] that we have need of a generic mechanism for "I want to bypass AR touching this value, but hand it untouched to the DB driver"), I'm now wondering if what we want is instead a more targeted solution to the original problem [that we've introduced edge-case-troublesome quoting for numerics], and have an `Arel.number` or similar, which will only accept numeric values. 🤔

Reply:
> how that original proposal helps Rational?

I'm not explaining myself clearly, and I'm mixing concepts here. I'm sorry. What I mean is that originally quoting _broke_ rational in that it added a `.to_s`. This breaks in the same way with the `.to_s`. See #44404.

I like this direction @matthewd. And you articulate better why the generic alias and and treatment of safe here I don't think is very good.

Let me take yet another stab at it, but this time with an exception simply for numerics.